src/PaletteStore.ts

```diff
@@ -0,0 +1,96 @@
+/*
+ * Copyright 2023 Inclusive Design Research Centre, OCAD University
+ * All rights reserved.
+ *
+ * Licensed under the New BSD license. You may not use this file except in
+ * compliance with this License.
+ *
+ * You may obtain a copy of the License at
+ * https://github.com/inclusive-design/adaptive-palette/blob/main/LICENSE
+ */
+
+"use strict";
+
+export class PaletteStore {
+
+  // Singleton storage for all palettes
+  // The contents are named Palette instances; hence, each palette must have
+  // a unique name.
+  static paletteMap = {};
+
+  /**
+   * Report if the PaletteStore is empty.
+   * @return: `true` if the store is empty; `false` otherwise.
+   */
+  isEmpty () {
+    return Object.keys(PaletteStore.paletteMap).length === 0;
+  }
+
+  /**
+   * Add a palette to the store, or replace a palette with a new one.  If the
+   * palette's name/identifier matches a palette already in the store, it
+   * replaces it.
+   * @param: {Palette} palette - The palette to add to the store.
+   * @param: {Palette}.name  - The internal name of the palette.
+   * @param: {String} name   - Optional, the preferred name of the palette.
+   */
+  addPalette (palette, paletteName?: string) {
```

Comment:
Add a type check for `palette` parameter.

Reply:
That requires a new type.  The only property that the `PaletteStore` uses is the name of the palette, for lookup.  A simple approach for a type then is:

```
type Palette = {
  name: string,
  palette: any
}
```

The problem is that lint will complain about the `any`.  Are we sure of the palette's structure?  If so, there are additional types to replace the `any`:

```
type Palette = {
  name: string,
  palette: JsonPalette  // In future, e.g., JsonPalette | OtherPalette | ...
}

type JsonPalette = {
  name: string,
  cells?: Cell
}

type Cell {
  type: string,
  options: Options,
}

type Options {
  label: string,
  bciAvId: number | (string|number)[],
  rowStart: number,
  rowSpan: number,
  columnStart: number,
  columnSpan: number
}

```

Note:  I'm not sure how to actually specify the `cells` field in the `JsonPalette` type.  It's a variable number of `Cell`s, but it is not an array.

What do think about the two approaches?  Should we use the complete one?

Reply:
I very much like the idea of a common place for shared types; very nice.

I had thought that this technique was not allowed because we were keeping code, styles, etc. grouped by the name of the component.  For example, there is a collection of  `ActionBmwCodeCell.ts`, `ActionBmwCodeCell.scss`, etc. files for everything that has to do with ActionBmwCodeCells.  That made me wonder how to share the types (and interfaces if we ever start to use them).

Breaking out the shared types into a single file makes total sense.  BTW, Where did the name `index.d.ts` come from?  Is that the practice with TypeScript?

Reply:
I've been reading the Preact repo to learn how it organizes the code base. `index.d.ts` is what I learnt there - https://github.com/preactjs/preact/blob/main/src/index.d.ts. Then I searched on `index.d.ts` and found a typescript doc about *.d.ts - https://www.typescriptlang.org/docs/handbook/declaration-files/templates/module-d-ts.html